# Introduction

In this notebook, we will do a comprehensive analysis of the Android app market by comparing thousands of apps in the Google Play store.

### Wrestle the Android App Store Data into Beautiful Looking Charts with Plotly

Have you ever thought about building your own an iOS or Android app? If so, then you probably have wondered about how things work in the app stores. Today we'll replicate some of the app store analytics provided by companies like App Annie or Sensor Tower that helps inform development and app marketing strategies for many companies. This stuff is BIG business!

**In this module, we will compare thousands of apps in the Google Play Store so that we can gain insight into:**

* How competitive different app categories (e.g., Games, Lifestyle, Weather) are

* Which app category offers compelling opportunities based on its popularity

* How many downloads you would give up by making your app paid vs. free

* How much you can reasonably charge for a paid app

* Which paid apps have had the highest revenue

* How many paid apps will recoup their development costs based on their sales revenue

<br>

**Today you'll learn:**

* How to quickly remove duplicates

* How to remove unwanted symbols and convert data into a numeric format

* How to wrangle columns containing nested data with Pandas

* How to create compelling data visualisations with the plotly library

* Create vertical, horizontal and grouped bar charts

* Create pie and donut charts for categorical data

* Use colour scales to make beautiful scatter plots

# About the Dataset of Google Play Store Apps & Reviews

**Data Source:** <br>
App and review data was scraped from the Google Play Store by Lavanya Gupta in 2018. Original files listed [here](
https://www.kaggle.com/lava18/google-play-store-apps).

# Import Statements

In [1]:
import pandas as pd


In [2]:
# prompt: mount from google drive

from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Notebook Presentation

In [3]:
# Show numeric output in decimal format e.g., 2.15
pd.options.display.float_format = '{:,.2f}'.format

# Read the Dataset

In [4]:
df_apps = pd.read_csv('/content/drive/MyDrive/jupyter/data/76/apps.csv')

# Data Cleaning

**Challenge**: How many rows and columns does `df_apps` have? What are the column names? Look at a random sample of 5 different rows with [.sample()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sample.html).

In [5]:
print(df_apps.shape)
print('\n')

(10841, 12)




In [6]:
df_apps.head()

App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  \
0        0      8.70        0  Paid  $13.99           Teen           Social   
1        0     33.00        0  Paid   $2.99       Everyone        Education   
2        0      5.50        0  Paid   $1.49       Everyone  Personalization   
3        0     19.00        0   NaN       0   Everyone 10+         Strategy   
4        0     10.00        0  Free       0       Everyone         Business   

     Last_Updated         Android_Ver  
0   July 28, 2017          4.1 and up  
1  April 15, 2016          3.0 and up  
2   July 11, 2018          4.2 and up  
3   June 28, 2018  Varies with device  
4  August 6, 2018          4.1 and up

We can already see that there are some data issues that we need to fix. In the Ratings and Type columns there are NaN (Not a number values) and in the Price column we have dollar signs that will cause problems.

In [7]:
df_apps.sample(4)

App Category  Rating   Reviews  Size_MBs  \
1518   BJ card game blackjack     GAME     NaN         3     21.00   
10767        Candy Crush Saga     GAME    4.40  22430188     74.00   
4713              BS Battery+    TOOLS    4.30       269      0.63   
2820       Ultimate Fighter Z     GAME    4.30       180     24.00   

          Installs  Type Price Content_Rating  Genres      Last_Updated  \
1518           500  Free     0           Teen    Card  December 2, 2016   
10767  500,000,000  Free     0       Everyone  Casual      July 5, 2018   
4713        50,000  Free     0       Everyone   Tools   August 18, 2012   
2820         5,000  Free     0       Everyone  Action     July 16, 2018   

      Android_Ver  
1518   2.3 and up  
10767  4.1 and up  
4713   2.0 and up  
2820   4.4 and up

### Drop Unused Columns

**Challenge**: Remove the columns called `Last_Updated` and `Android_Version` from the DataFrame. We will not use these columns.

In [8]:
# we will remove columns that we are not going to use (make the dataset smaller)
df_apps.drop(['Last_Updated', 'Android_Ver'], axis=1, inplace=True)

In [9]:
df_apps.sample(3)

App      Category  Rating  Reviews  Size_MBs  \
557                    EB Scanner  PRODUCTIVITY    5.00        9     25.00   
8598  Bus Driver 3D: Hill Station        FAMILY    4.30    31705     49.00   
297              FK Sloboda Tuzla        SPORTS     NaN        0     26.00   

       Installs  Type Price Content_Rating        Genres  
557          50  Free     0       Everyone  Productivity  
8598  5,000,000  Free     0       Everyone    Simulation  
297          10  Free     0       Everyone        Sports

### Find and Remove NaN values in Ratings

**Challenge**: How may rows have a NaN value (not-a-number) in the Ratings column? Create DataFrame called `df_apps_clean` that does not include these rows.

In [10]:
# get some info from the original dataframe
print(df_apps.shape)
print('\n')
df_apps.info()

(10841, 10)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  int64  
 4   Size_MBs        10841 non-null  float64
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content_Rating  10841 non-null  object 
 9   Genres          10841 non-null  object 
dtypes: float64(2), int64(1), object(7)
memory usage: 847.1+ KB


In [11]:
# get the rows with Nan values in the rating
nan_rows = df_apps[df_apps.Rating.isna()]
print(nan_rows.shape)
print('\n')
nan_rows.head()


(1474, 10)




App         Category  Rating  \
0                       Ak Parti Yardım Toplama           SOCIAL     NaN   
1                    Ain Arabic Kids Alif Ba ta           FAMILY     NaN   
2  Popsicle Launcher for Android P 9.0 launcher  PERSONALIZATION     NaN   
3                     Command & Conquer: Rivals           FAMILY     NaN   
4                                    CX Network         BUSINESS     NaN   

   Reviews  Size_MBs Installs  Type   Price Content_Rating           Genres  
0        0      8.70        0  Paid  $13.99           Teen           Social  
1        0     33.00        0  Paid   $2.99       Everyone        Education  
2        0      5.50        0  Paid   $1.49       Everyone  Personalization  
3        0     19.00        0   NaN       0   Everyone 10+         Strategy  
4        0     10.00        0  Free       0       Everyone         Business

In [12]:
# from the dataframe above, you can see that NaN ratings means 0 reviews.
# we are not interested in apps that have no ratings (that is not useful data) so we drop it
df_apps_clean = df_apps.dropna()
df_apps_clean.shape

(9367, 10)

### Find and Remove Duplicates

**Challenge**: Are there any duplicates in data? Check for duplicates using the [.duplicated()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.duplicated.html) function. How many entries can you find for the "Instagram" app? Use [.drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) to remove any duplicates from `df_apps_clean`.


**Find the number of duplicates**

In [13]:
# check for duplicates in the dataframe (whole)
duplicates = df_apps_clean.duplicated(subset=None, keep='first')  # subset=None by default (check duplicates for all columns), 'first' means set True for the first duplicate
df_duplicates = df_apps_clean[duplicates]
df_duplicates

App            Category  Rating  \
946                        420 BZ Budeze Delivery             MEDICAL    5.00   
1133                                  MouseMingle              DATING    2.70   
1196   Cardiac diagnosis (heart rate, arrhythmia)             MEDICAL    4.40   
1231                                 Sway Medical             MEDICAL    5.00   
1247               Chat Kids - Chat Room For Kids              DATING    4.70   
...                                           ...                 ...     ...   
10802               Skype - free IM & video calls       COMMUNICATION    4.10   
10809                                   Instagram              SOCIAL    4.50   
10826                                Google Drive        PRODUCTIVITY    4.40   
10832                                 Google News  NEWS_AND_MAGAZINES    3.90   
10839                              Subway Surfers                GAME    4.50   

        Reviews  Size_MBs       Installs  Type   Price Content_Rating  \
946           2     11.00            100  Free       0     Mature 17+   
1133          3      3.90            100  Free       0     Mature 17+   
1196          8      6.50            100  Paid  $12.99       Everyone   
1231          3     22.00            100  Free       0       Everyone   
1247          6      4.90            100  Free       0     Mature 17+   
...         ...       ...            ...   ...     ...            ...   
10802  10484169      3.50  1,000,000,000  Free       0       Everyone   
10809  66577313      5.30  1,000,000,000  Free       0           Teen   
10826   2731211      4.00  1,000,000,000  Free       0       Everyone   
10832    877635     13.00  1,000,000,000  Free       0           Teen   
10839  27725352     76.00  1,000,000,000  Free       0   Everyone 10+   

                 Genres  
946             Medical  
1133             Dating  
1196            Medical  
1231            Medical  
1247             Dating  
...                 ...  
10802     Communication  
10809            Social  
10826      Productivity  
10832  News & Magazines  
10839            Arcade  

[476 rows x 10 columns]

In [14]:
# set True to all duplicates with keep=False (don't exclude the first one and keep it in the dataframe)
columns = df_apps_clean.columns
duplicates_v2 = df_apps_clean[df_apps_clean.duplicated(subset=None, keep=False)]
print(duplicates_v2.shape)    # there are many more values because we get all duplicates (we don't exclude the 'first' like before)

(879, 10)


In [15]:
print(f"Duplicates excluding the 'first' intance: {df_duplicates.shape}")
print(f"All duplicate instances: {duplicates_v2.shape}")

Duplicates excluding the 'first' intance: (476, 10)
All duplicate instances: (879, 10)


In [16]:
# count the number of instances for each duplicate row
duplicate_counts = duplicates_v2.groupby(list(duplicates_v2.columns)).size()
print(type(duplicate_counts))
duplicate_counts

<class 'pandas.core.series.Series'>


App                                                 Category            Rating  Reviews  Size_MBs  Installs     Type  Price   Content_Rating  Genres          
10 Best Foods for You                               HEALTH_AND_FITNESS  4.00    2490     3.80      500,000      Free  0       Everyone 10+    Health & Fitness    2
1800 Contacts - Lens Store                          MEDICAL             4.70    23160    26.00     1,000,000    Free  0       Everyone        Medical             2
2017 EMRA Antibiotic Guide                          MEDICAL             4.40    12       3.80      1,000        Paid  $16.99  Everyone        Medical             2
21-Day Meditation Experience                        HEALTH_AND_FITNESS  4.40    11506    15.00     100,000      Free  0       Everyone        Health & Fitness    2
420 BZ Budeze Delivery                              MEDICAL             5.00    2        11.00     100          Free  0       Mature 17+      Medical             2
                     

In [17]:
# convert the series into a dataframe for better readability
duplicate_counts_df = duplicate_counts.reset_index(name='count')
duplicate_counts_df.head()    # remove head to sort using interactive table

App            Category  Rating  Reviews  \
0         10 Best Foods for You  HEALTH_AND_FITNESS    4.00     2490   
1    1800 Contacts - Lens Store             MEDICAL    4.70    23160   
2    2017 EMRA Antibiotic Guide             MEDICAL    4.40       12   
3  21-Day Meditation Experience  HEALTH_AND_FITNESS    4.40    11506   
4        420 BZ Budeze Delivery             MEDICAL    5.00        2   

   Size_MBs   Installs  Type   Price Content_Rating            Genres  count  
0      3.80    500,000  Free       0   Everyone 10+  Health & Fitness      2  
1     26.00  1,000,000  Free       0       Everyone           Medical      2  
2      3.80      1,000  Paid  $16.99       Everyone           Medical      2  
3     15.00    100,000  Free       0       Everyone  Health & Fitness      2  
4     11.00        100  Free       0     Mature 17+           Medical      2

In [18]:
# remove the duplicates that you see in the columns (App, Type, Price)
# the subset to identify duplicates will be (App, Type, Price)
df_apps_clean =df_apps_clean.drop_duplicates(subset=['App', 'Type', 'Price'])
df_apps_clean[df_apps_clean.App == 'Instagram']

App Category  Rating   Reviews  Size_MBs       Installs  Type  \
10806  Instagram   SOCIAL    4.50  66577313      5.30  1,000,000,000  Free   

      Price Content_Rating  Genres  
10806     0           Teen  Social

In [19]:
#

### What else should I know about the data?


So we can see that 13 different features were originally scraped from the Google Play Store.


* Obviously, the data is just a sample out of all the Android apps. It doesn't include all Android apps of which there are millions.

* I’ll assume that the sample is representative of the App Store as a whole. This is not necessarily the case as, during the web scraping process, this sample was served up based on geographical location and user behaviour of the person who scraped it - in our case Lavanya Gupta.

* The data was compiled around 2017/2018. The pricing data reflect the price in USD Dollars at the time of scraping. (developers can offer promotions and change their app’s pricing).

* I’ve converted the app’s size to a floating-point number in MBs. If data was missing, it has been replaced by the average size for that category.

* The installs are not the exact number of installs. If an app has 245,239 installs then Google will simply report an order of magnitude like 100,000+. I’ve removed the '+' and we’ll assume the exact number of installs in that column for simplicity.




# Find Highest Rated Apps

**Challenge**: Identify which apps are the highest rated. What problem might you encounter if you rely exclusively on ratings alone to determine the quality of an app?

In [20]:
df_apps_clean.sort_values('Rating', ascending=False).head()

App     Category  Rating  Reviews  Size_MBs Installs  \
21    KBA-EZ Health Guide      MEDICAL    5.00        4     25.00        1   
1230         Sway Medical      MEDICAL    5.00        3     22.00      100   
1227    AJ Men's Grooming    LIFESTYLE    5.00        2     22.00      100   
1224       FK Dedinje BGD       SPORTS    5.00       36      2.60      100   
1223      CB VIDEO VISION  PHOTOGRAPHY    5.00       13      2.60      100   

      Type Price Content_Rating       Genres  
21    Free     0       Everyone      Medical  
1230  Free     0       Everyone      Medical  
1227  Free     0       Everyone    Lifestyle  
1224  Free     0       Everyone       Sports  
1223  Free     0       Everyone  Photography

# Find 5 Largest Apps in terms of Size (MBs)

**Challenge**: What's the size in megabytes (MB) of the largest Android apps in the Google Play Store. Based on the data, do you think there could be limit in place or can developers make apps as large as they please?

In [21]:
df_apps_clean.sort_values('Size_MBs', ascending=False).head()

App            Category  Rating  Reviews  \
9942   Talking Babsy Baby: Baby Games           LIFESTYLE    4.00   140995   
10687          Hungry Shark Evolution                GAME    4.50  6074334   
9943            Miami crime simulator                GAME    4.00   254518   
9944     Gangster Town: Vice District              FAMILY    4.30    65146   
3144                       Vi Trainer  HEALTH_AND_FITNESS    3.60      124   

       Size_MBs     Installs  Type Price Content_Rating  \
9942     100.00   10,000,000  Free     0       Everyone   
10687    100.00  100,000,000  Free     0           Teen   
9943     100.00   10,000,000  Free     0     Mature 17+   
9944     100.00   10,000,000  Free     0     Mature 17+   
3144     100.00        5,000  Free     0       Everyone   

                       Genres  
9942   Lifestyle;Pretend Play  
10687                  Arcade  
9943                   Action  
9944               Simulation  
3144         Health & Fitness

# Find the 5 App with Most Reviews

**Challenge**: Which apps have the highest number of reviews? Are there any paid apps among the top 50?

In [22]:
df_apps_clean.sort_values('Reviews', ascending=False).head(5)

App       Category  Rating  \
10805                                  Facebook         SOCIAL    4.10   
10785                        WhatsApp Messenger  COMMUNICATION    4.40   
10806                                 Instagram         SOCIAL    4.50   
10784  Messenger – Text and Video Chat for Free  COMMUNICATION    4.00   
10650                            Clash of Clans           GAME    4.60   

        Reviews  Size_MBs       Installs  Type Price Content_Rating  \
10805  78158306      5.30  1,000,000,000  Free     0           Teen   
10785  69119316      3.50  1,000,000,000  Free     0       Everyone   
10806  66577313      5.30  1,000,000,000  Free     0           Teen   
10784  56642847      3.50  1,000,000,000  Free     0       Everyone   
10650  44891723     98.00    100,000,000  Free     0   Everyone 10+   

              Genres  
10805         Social  
10785  Communication  
10806         Social  
10784  Communication  
10650       Strategy

If you look at the number of reviews, you can find the most popular apps on the Android App Store. These include the usual suspects: Facebook, WhatsApp, Instagram etc. What’s also notable is that the list of the top 50 most reviewed apps does not include a single paid app

<br>

Here we can clearly see that there seems to be an upper bound of 100 MB for the size of an app. A quick google search would also have revealed that this limit is imposed by the Google Play Store itself. It’s interesting to see that a number of apps actually hit that limit exactly.

# Plotly Pie and Donut Charts - Visualise Categorical Data: Content Ratings

In [23]:
# find the distribution of the Content rating
ratings = df_apps_clean.Content_Rating.value_counts()
ratings

Everyone           6621
Teen                912
Mature 17+          357
Everyone 10+        305
Adults only 18+       3
Unrated               1
Name: Content_Rating, dtype: int64

In [24]:
import plotly.express as px

In [25]:
# create a pie chart
fig = px.pie(labels=ratings.index, values=ratings.values)
fig.show()

/usr/local/lib/python3.10/dist-packages/plotly/express/_core.py:137: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  return args["labels"][column]


In [26]:
import numpy as np

In [27]:
# resolving for the warning on plotly
# need to convert the Series into a numpy array for appropriate indexing
fig = px.pie(labels=np.array(ratings.index), values=np.array(ratings.values))
fig.show()

In [28]:
fig = px.pie(labels=np.array(ratings.index),
             values=np.array(ratings.values),
             title='Content Rating',
             names=ratings.index)

# traces -- refers to the graphical marks on a figure. Think of traces as a collection of attributes
fig.update_traces(textposition='outside', textinfo='percent+label')

fig.show()

### Create a Donut Chart

In [29]:
fig = px.pie(labels=np.array(ratings.index),
             values=np.array(ratings.values),
             title='Content Rating',
             names=ratings.index,
             hole=0.5)

# traces -- refers to the graphical marks on a figure. Think of traces as a collection of attributes
fig.update_traces(textposition='inside',
                  textinfo='percent',
                  textfont_size=14)

fig.show()

# Numeric Type Conversion: Examine the Number of Installs

**Challenge**: How many apps had over 1 billion (that's right - BILLION) installations? How many apps just had a single install?

Check the datatype of the Installs column.

Count the number of apps at each level of installations.

Convert the number of installations (the Installs column) to a numeric data type. Hint: this is a 2-step process. You'll have make sure you remove non-numeric characters first.

In [30]:
df_apps_clean.Installs.describe()

count          8199
unique           19
top       1,000,000
freq           1417
Name: Installs, dtype: object

In [31]:
# useless information due to Installs being a string (object)
df_apps_clean[['App', 'Installs']].groupby('Installs').count()

App
Installs           
1                 3
1,000           698
1,000,000      1417
1,000,000,000    20
10               69
10,000          988
10,000,000      933
100             303
100,000        1096
100,000,000     189
5                 9
5,000           425
5,000,000       607
50               56
50,000          457
50,000,000      202
500             199
500,000         504
500,000,000      24

In [32]:
# remove the commas from the numbers (string)
df_apps_clean.Installs.astype(str).str.replace(',', '')

21                1
28                1
47                1
82                5
99                5
            ...    
10824    1000000000
10828    1000000000
10829    1000000000
10831    1000000000
10835    1000000000
Name: Installs, Length: 8199, dtype: object

In [33]:
# ----- A BETTER WAY --------------
new_column = df_apps_clean.Installs.str.replace(',', '').astype(int)
print(new_column)
df_apps_clean.Installs = new_column

21                1
28                1
47                1
82                5
99                5
            ...    
10824    1000000000
10828    1000000000
10829    1000000000
10831    1000000000
10835    1000000000
Name: Installs, Length: 8199, dtype: int64


In [34]:
df_apps_clean[['App', 'Installs']].groupby('Installs').count()

App
Installs        
1              3
5              9
10            69
50            56
100          303
500          199
1000         698
5000         425
10000        988
50000        457
100000      1096
500000       504
1000000     1417
5000000      607
10000000     933
50000000     202
100000000    189
500000000     24
1000000000    20

# Find the Most Expensive Apps, Filter out the Junk, and Calculate a (ballpark) Sales Revenue Estimate

Let's examine the Price column more closely.

**Challenge**: Convert the price column to numeric data. Then investigate the top 20 most expensive apps in the dataset.

Remove all apps that cost more than $250 from the `df_apps_clean` DataFrame.

Add a column called 'Revenue_Estimate' to the DataFrame. This column should hold the price of the app times the number of installs. What are the top 10 highest grossing paid apps according to this estimate? Out of the top 10 highest grossing paid apps, how many are games?


In [35]:
df_apps_clean.Price = df_apps_clean.Price.str.replace('$', '').astype(float)
df_apps_clean.Price

<ipython-input-35-bb4f2ce0bf61>:1: FutureWarning:

The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.



21      0.00
28      1.49
47      0.99
82      0.00
99      0.00
        ... 
10824   0.00
10828   0.00
10829   0.00
10831   0.00
10835   0.00
Name: Price, Length: 8199, dtype: float64

In [36]:
df_apps_clean.sort_values('Price', ascending=False).head(10)

App   Category  Rating  Reviews  Size_MBs  \
3946        I'm Rich - Trump Edition  LIFESTYLE    3.60      275      7.30   
2461              I AM RICH PRO PLUS    FINANCE    4.00       36     41.00   
4606               I Am Rich Premium    FINANCE    4.10     1867      4.70   
3145              I am rich(premium)    FINANCE    3.50      472      0.94   
3554                      💎 I'm rich  LIFESTYLE    3.80      718     26.00   
5765                       I am rich  LIFESTYLE    3.80     3547      1.80   
1946  I am rich (Most expensive app)    FINANCE    4.10      129      2.70   
2775                   I Am Rich Pro     FAMILY    4.40      201      2.70   
3221                  I am Rich Plus     FAMILY    4.00      856      8.70   
3114                       I am Rich    FINANCE    4.30      180      3.80   

      Installs  Type  Price Content_Rating         Genres  
3946     10000  Paid 400.00       Everyone      Lifestyle  
2461      1000  Paid 399.99       Everyone        Finance  
4606     50000  Paid 399.99       Everyone        Finance  
3145      5000  Paid 399.99       Everyone        Finance  
3554     10000  Paid 399.99       Everyone      Lifestyle  
5765    100000  Paid 399.99       Everyone      Lifestyle  
1946      1000  Paid 399.99           Teen        Finance  
2775      5000  Paid 399.99       Everyone  Entertainment  
3221     10000  Paid 399.99       Everyone  Entertainment  
3114      5000  Paid 399.99       Everyone        Finance

What’s going on here? There are 15 I am Rich Apps in the Google Play Store apparently. They all cost `$300` or more, which is the main point of the app. The story goes that in 2008, Armin Heinrich released the very first I am Rich app in the iOS App Store for $999.90. The app does absolutely nothing. It just displays the picture of a gemstone and can be used to prove to your friends how rich you are. Armin actually made a total of 7 sales before the app was hastily removed by Apple. Nonetheless, it inspired a bunch of copycats on the Android App Store, but if you search today, you’ll find all of these apps have disappeared as well. The high installation numbers are likely gamed by making the app was available for free at some point to get reviews and appear more legitimate.

### The most expensive apps sub $250

In [37]:
# we have removed any app above the $250 range
df_apps_clean = df_apps_clean[df_apps_clean.Price < 250]
df_apps_clean.sort_values('Price', ascending=False).head(5)

App   Category  Rating  Reviews  Size_MBs  \
2281  Vargo Anesthesia Mega App    MEDICAL    4.60       92     32.00   
1407               LTC AS Legal    MEDICAL    4.00        6      1.30   
2629           I am Rich Person  LIFESTYLE    4.20      134      1.80   
2481    A Manual of Acupuncture    MEDICAL    3.50      214     68.00   
2463         PTA Content Master    MEDICAL    4.20       64     41.00   

      Installs  Type  Price Content_Rating     Genres  
2281      1000  Paid  79.99       Everyone    Medical  
1407       100  Paid  39.99       Everyone    Medical  
2629      1000  Paid  37.99       Everyone  Lifestyle  
2481      1000  Paid  33.99       Everyone    Medical  
2463      1000  Paid  29.99       Everyone    Medical

### Highest Grossing Paid Apps (ballpark estimate)

In [38]:
# get the highest grossing paid apps (installs * price)
# multiple intalls and price using 'mul()'
df_apps_clean['Revenue Estimate'] = df_apps_clean.Installs.mul(df_apps_clean.Price)
df_apps_clean.sort_values('Revenue Estimate', ascending=False).head(10)

<ipython-input-38-2daf13033447>:3: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



App     Category  Rating  Reviews  Size_MBs  \
9220                      Minecraft       FAMILY    4.50  2376564     19.00   
8825                  Hitman Sniper         GAME    4.60   408292     29.00   
7151  Grand Theft Auto: San Andreas         GAME    4.40   348962     26.00   
7477            Facetune - For Free  PHOTOGRAPHY    4.40    49553     48.00   
7977        Sleep as Android Unlock    LIFESTYLE    4.50    23966      0.85   
6594            DraStic DS Emulator         GAME    4.60    87766     12.00   
6082                   Weather Live      WEATHER    4.50    76593      4.75   
7954                    Bloons TD 5       FAMILY    4.60   190086     94.00   
7633        Five Nights at Freddy's         GAME    4.60   100805     50.00   
6746     Card Wars - Adventure Time       FAMILY    4.30   129603     23.00   

      Installs  Type  Price Content_Rating                     Genres  \
9220  10000000  Paid   6.99   Everyone 10+  Arcade;Action & Adventure   
8825  10000000  Paid   0.99     Mature 17+                     Action   
7151   1000000  Paid   6.99     Mature 17+                     Action   
7477   1000000  Paid   5.99       Everyone                Photography   
7977   1000000  Paid   5.99       Everyone                  Lifestyle   
6594   1000000  Paid   4.99       Everyone                     Action   
6082    500000  Paid   5.99       Everyone                    Weather   
7954   1000000  Paid   2.99       Everyone                   Strategy   
7633   1000000  Paid   2.99           Teen                     Action   
6746   1000000  Paid   2.99   Everyone 10+    Card;Action & Adventure   

      Revenue Estimate  
9220     69,900,000.00  
8825      9,900,000.00  
7151      6,990,000.00  
7477      5,990,000.00  
7977      5,990,000.00  
6594      4,990,000.00  
6082      2,995,000.00  
7954      2,990,000.00  
7633      2,990,000.00  
6746      2,990,000.00

The top spot of the highest-grossing paid app goes to … Minecraft at close to $70 million. It’s quite interesting that Minecraft (along with Bloons and Card Wars) is actually listed in the Family category rather than in the Game category. If we include these titles, we see that 7 out the top 10 highest-grossing apps are games. The Google Play Store seems to be quite flexible with its category labels.

# Plotly Bar Charts & Scatter Plots: Analysing App Categories

If you were to release an app, would you choose to go after a competitive category with many other apps? Or would you target a popular category with a high number of downloads? Or perhaps you can target a category which is both popular but also one where the downloads are spread out among many different apps. That way, even if it’s more difficult to discover among all the other apps, your app has a better chance of getting installed, right? Let’s analyse this with bar charts and scatter plots and figure out which categories are dominating the market.

In [39]:
# get number of different categories
df_apps_clean.Category.nunique()

33

In [40]:
top10_category = df_apps_clean.Category.value_counts()[:10]
top10_category

FAMILY             1606
GAME                910
TOOLS               719
PRODUCTIVITY        301
PERSONALIZATION     298
LIFESTYLE           297
FINANCE             296
MEDICAL             292
PHOTOGRAPHY         263
BUSINESS            262
Name: Category, dtype: int64

### Vertical Bar Chart - Highest Competition (Number of Apps)

In [41]:
bar = px.bar(x=top10_category.index,
             y=top10_category.values)

bar.show()

### Horizontal Bar Chart - Most Popular Categories (Highest Downloads)

But what if we look at it from a different perspective? What matters is not just the total number of apps in the category but how often apps are downloaded in that category. This will give us an idea of how popular a category is. First, we have to group all our apps by category and sum the number of installations:

In [42]:
category_installs = df_apps_clean.groupby('Category').agg({'Installs': pd.Series.sum})
category_installs.sort_values('Installs', ascending=True, inplace=True)

In [43]:
category_installs.head()

Installs
Category           
EVENTS     15949410
BEAUTY     26916200
PARENTING  31116110
MEDICAL    39162676
COMICS     44931100

In [44]:
h_bar = px.bar(x = category_installs.Installs,
                   y = category_installs.index,
                   orientation='h')

h_bar.show()

In [45]:
# add title
h_bar = px.bar(x = category_installs.Installs,
                   y = category_installs.index,
                   orientation='h',
                   title='Category Popularity')

h_bar.update_layout(xaxis_title='Number of Downloads', yaxis_title='Category')
h_bar.show()

### Category Concentration - Downloads vs. Competition

**Challenge**:
* First, create a DataFrame that has the number of apps in one column and the number of installs in another:

<img src=https://imgur.com/uQRSlXi.png width="350">

* Then use the [plotly express examples from the documentation](https://plotly.com/python/line-and-scatter/) alongside the [.scatter() API reference](https://plotly.com/python-api-reference/generated/plotly.express.scatter.html)to create scatter plot that looks like this.

<img src=https://imgur.com/cHsqh6a.png>

*Hint*: Use the size, hover_name and color parameters in .scatter(). To scale the yaxis, call .update_layout() and specify that the yaxis should be on a log-scale like so: yaxis=dict(type='log')

In [46]:
category_df = df_apps_clean.groupby('Category').agg({
    'App': 'count',
    'Installs': 'sum'
})
category_df.sort_values('App', ascending=False)

App     Installs
Category                              
FAMILY               1606   4437554490
GAME                  910  13858762717
TOOLS                 719   8099724500
PRODUCTIVITY          301   5788070180
PERSONALIZATION       298   1532352930
LIFESTYLE             297    503611120
FINANCE               296    455249400
MEDICAL               292     39162676
PHOTOGRAPHY           263   4649143130
BUSINESS              262    692018120
SPORTS                260   1096431465
COMMUNICATION         257  11039241530
HEALTH_AND_FITNESS    243   1134006220
NEWS_AND_MAGAZINES    204   2369110650
SOCIAL                203   5487841475
TRAVEL_AND_LOCAL      187   2894859300
SHOPPING              180   1400331540
BOOKS_AND_REFERENCE   169   1665791655
VIDEO_PLAYERS         148   3916897200
DATING                134    140912410
EDUCATION             118    352852000
MAPS_AND_NAVIGATION   118    503267560
ENTERTAINMENT         102   2113660000
FOOD_AND_DRINK         94    211677750
AUTO_AND_VEHICLES      73     53129800
WEATHER                72    361096500
LIBRARIES_AND_DEMO     64     52083000
HOUSE_AND_HOME         62     97082000
ART_AND_DESIGN         61    114233100
COMICS                 54     44931100
PARENTING              50     31116110
EVENTS                 45     15949410
BEAUTY                 42     26916200

In [47]:
# create scatter plot
scatter = px.scatter(category_df, # data
                    x='App', # column name
                    y='Installs',
                    title='Category Concentration',
                    size='App',
                    hover_name=category_df.index,
                    color='Installs')

scatter.update_layout(xaxis_title="Number of Apps (Lower=More Concentrated)",
                      yaxis_title="Installs",
                      yaxis=dict(type='log')) # y-axis as log-scale

scatter.show()

What we see is that the categories like Family, Tools, and Game have many different apps sharing a high number of downloads. But for the categories like video players and entertainment, all the downloads are concentrated in very few apps.

# Extracting Nested Data from a Column using .stack()

**Challenge**: How many different types of genres are there? Can an app belong to more than one genre? Check what happens when you use .value_counts() on a column with nested values? See if you can work around this problem by using the .split() function and the DataFrame's [.stack() method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.stack.html).


The `stack` method in pandas is used for reshaping a DataFrame. When you stack a DataFrame, you are pivoting the columns into the rows, essentially "compressing" a level in the DataFrame's columns to produce a longer, narrower DataFrame. It's a part of pandas' multi-level indexing or hierarchical indexing feature.


**Stack method basically skinnies a dataframe (makes it longer in rows, and shorter in columns). It 'compresses' a dataframe and creates multi-level index (e.g. instead of index = 0, we have index = [0, A])**

In [48]:
df_apps_clean.head(5)

App Category  Rating  Reviews  \
21                              KBA-EZ Health Guide  MEDICAL    5.00        4   
28                                         Ra Ga Ba     GAME    5.00        2   
47                                          Mu.F.O.     GAME    5.00        2   
82                                 Brick Breaker BR     GAME    5.00        7   
99  Anatomy & Physiology Vocabulary Exam Review App  MEDICAL    5.00        1   

    Size_MBs  Installs  Type  Price Content_Rating   Genres  Revenue Estimate  
21     25.00         1  Free   0.00       Everyone  Medical              0.00  
28     20.00         1  Paid   1.49       Everyone   Arcade              1.49  
47     16.00         1  Paid   0.99       Everyone   Arcade              0.99  
82     19.00         5  Free   0.00       Everyone   Arcade              0.00  
99      4.60         5  Free   0.00       Everyone  Medical              0.00

In [49]:
# how many types of genres
df_apps_clean.value_counts('Genres')

Genres
Tools                          718
Entertainment                  467
Education                      429
Productivity                   301
Personalization                298
                              ... 
Board;Pretend Play               1
Lifestyle;Pretend Play           1
Music & Audio;Music & Video      1
Parenting;Brain Games            1
Comics;Creativity                1
Length: 114, dtype: int64

In [50]:
# you can see that the last genres with just 1 value are sometimes separated with ;
# we need to separate the genre names
df_apps_clean.Genres.nunique()

114

In [51]:
# separate genre from semi-colon ;
# stack method basically removes the newly created column
stack = df_apps_clean.Genres.str.split(';', expand=True)
print(f'Unstacked: {stack}')
stack = stack.stack()
print(f'We now have a single column with shape: {stack.shape}')
stack

Unstacked:                              0     1
21                     Medical  None
28                      Arcade  None
47                      Arcade  None
82                      Arcade  None
99                     Medical  None
...                        ...   ...
10824             Productivity  None
10828  Video Players & Editors  None
10829  Video Players & Editors  None
10831         News & Magazines  None
10835                   Arcade  None

[8184 rows x 2 columns]
We now have a single column with shape: (8564,)


21     0                    Medical
28     0                     Arcade
47     0                     Arcade
82     0                     Arcade
99     0                    Medical
                     ...           
10824  0               Productivity
10828  0    Video Players & Editors
10829  0    Video Players & Editors
10831  0           News & Magazines
10835  0                     Arcade
Length: 8564, dtype: object

In [52]:
num_genres = stack.value_counts()
print(f'Number of genres: {len(num_genres)}')

Number of genres: 53


In [56]:
# after .stack(), the first element becomes a multi-level index (tuple of (21, 0))
# the column was 'compressed' into a larger row
stack.index[0]

(21, 0)

# Colour Scales in Plotly Charts - Competition in Genres

**Challenge**: Can you create this chart with the Series containing the genre data?

<img src=https://imgur.com/DbcoQli.png width=400>

Try experimenting with the built in colour scales in Plotly. You can find a full list [here](https://plotly.com/python/builtin-colorscales/).

* Find a way to set the colour scale using the color_continuous_scale parameter.
* Find a way to make the color axis disappear by using coloraxis_showscale.

In [53]:
bar = px.bar(x=num_genres.index[:10],
             y=num_genres.values[:10],
             title="Most Apps per Genre",
             hover_name=num_genres.index[:10],  # provide the genre name when you hover over the bars
             color=num_genres.index[:10],   # provide the values that you want colored
             color_continuous_scale='agsunset') # decide on a color scale (predefined)

bar.show()

# Grouped Bar Charts: Free vs. Paid Apps per Category

In [57]:
df_apps_clean.head()

App Category  Rating  Reviews  \
21                              KBA-EZ Health Guide  MEDICAL    5.00        4   
28                                         Ra Ga Ba     GAME    5.00        2   
47                                          Mu.F.O.     GAME    5.00        2   
82                                 Brick Breaker BR     GAME    5.00        7   
99  Anatomy & Physiology Vocabulary Exam Review App  MEDICAL    5.00        1   

    Size_MBs  Installs  Type  Price Content_Rating   Genres  Revenue Estimate  
21     25.00         1  Free   0.00       Everyone  Medical              0.00  
28     20.00         1  Paid   1.49       Everyone   Arcade              1.49  
47     16.00         1  Paid   0.99       Everyone   Arcade              0.99  
82     19.00         5  Free   0.00       Everyone   Arcade              0.00  
99      4.60         5  Free   0.00       Everyone  Medical              0.00

In [59]:
# you can groupby multiple columns, these become indexes (multi-level)
df_free_vs_paid = df_apps_clean.groupby(['Category', 'Type']).agg({'App': 'count'})
df_free_vs_paid.head()

App
Category          Type     
ART_AND_DESIGN    Free   58
                  Paid    3
AUTO_AND_VEHICLES Free   72
                  Paid    1
BEAUTY            Free   42

In [60]:
# you can avoid creating multi-level indexes by specifying it in the groupby function
df_free_vs_paid = df_apps_clean.groupby(['Category', 'Type'], as_index=False).agg({'App': 'count'})
df_free_vs_paid.head()  # now you have the groupby columns as columns, not index

Category  Type  App
0     ART_AND_DESIGN  Free   58
1     ART_AND_DESIGN  Paid    3
2  AUTO_AND_VEHICLES  Free   72
3  AUTO_AND_VEHICLES  Paid    1
4             BEAUTY  Free   42

In [62]:
# the most popular apps are FAMILY category and Free
df_free_vs_paid.sort_values('App', ascending=False)

Category  Type   App
19              FAMILY  Free  1456
25                GAME  Free   834
53               TOOLS  Free   656
21             FINANCE  Free   289
31           LIFESTYLE  Free   284
..                 ...   ...   ...
17       ENTERTAINMENT  Paid     2
24      FOOD_AND_DRINK  Paid     2
40           PARENTING  Paid     2
38  NEWS_AND_MAGAZINES  Paid     2
3    AUTO_AND_VEHICLES  Paid     1

[61 rows x 3 columns]

**Challenge**: Use the plotly express bar [chart examples](https://plotly.com/python/bar-charts/#bar-chart-with-sorted-or-ordered-categories) and the [.bar() API reference](https://plotly.com/python-api-reference/generated/plotly.express.bar.html#plotly.express.bar) to create this bar chart:

<img src=https://imgur.com/LE0XCxA.png>

You'll want to use the `df_free_vs_paid` DataFrame that you created above that has the total number of free and paid apps per category.

See if you can figure out how to get the look above by changing the `categoryorder` to 'total descending' as outlined in the documentation here [here](https://plotly.com/python/categorical-axes/#automatically-sorting-categories-by-name-or-total-value).

In [63]:
bar = px.bar(x=df_free_vs_paid.Category,
             y=df_free_vs_paid.App,
             barmode='group',
             labels=df_free_vs_paid.Type.unique(),
             title='Free vs Paid Apps by Category')

bar.update_traces()

bar.show()


In [66]:
"""
Color='Type' determines what column to color (sometimes you need to input the
dataframe or it won't recognize the column). Like in this example.
barmode='group' -- tells it to have a bar side by side
"""
bar = px.bar(data_frame=df_free_vs_paid,
             x=df_free_vs_paid.Category,
             y=df_free_vs_paid.App,
             title='Free vs Paid Apps by Category',
             color='Type',
             barmode='group')

# sets the xaxis and yaxis titles
# xaxis - https://plotly.com/python/categorical-axes/#automatically-sorting-categories-by-name-or-total-value
# xaxis - you can also use the fig.update_xaxes(categoryorder='total descending')
# yaxis - setting it to type log for a smoothing function
bar.update_layout(xaxis_title='Category',
                  yaxis_title='Number of Apps',
                  xaxis={'categoryorder': 'total descending'},
                  yaxis=dict(type='log'))

bar.show()


In [67]:
t = dict(type='log')

{'type': 'log'}

What we see is that while there are very few paid apps on the Google Play Store, some categories have relatively more paid apps than others, including Personalization, Medical and Weather. So, depending on the category you are targeting, it might make sense to release a paid-for app

# Plotly Box Plots: Lost Downloads for Paid Apps

**Challenge**: Create a box plot that shows the number of Installs for free versus paid apps. How does the median number of installations compare? Is the difference large or small?

Use the [Box Plots Guide](https://plotly.com/python/box-plots/) and the [.box API reference](https://plotly.com/python-api-reference/generated/plotly.express.box.html) to create the following chart.

<img src=https://imgur.com/uVsECT3.png>


## Box Plot Explanation

The graph you have shared is a box plot, which is also known as a box-and-whisker plot. It's used to display the distribution of a dataset and is particularly useful for showing median values, quartiles, and outliers. Here's how to read it:

1. **Type of Data**: The x-axis shows two categories of apps, "Free" and "Paid". These represent the types of apps in the dataset.

2. **Number of Installs**: The y-axis is on a logarithmic scale and represents the number of installs that apps have received. It spans from 1 to 100 million installs, indicating a wide range in the number of installs.

3. **Box**: For each category (Free and Paid), there's a box that represents the interquartile range (IQR):
   - The bottom of the box represents the first quartile (Q1), which is the 25th percentile.
   - The line in the middle of the box is the median (Q2), the 50th percentile.
   - The top of the box represents the third quartile (Q3), which is the 75th percentile.
   
   The distance between the top and bottom of the box (the IQR) shows the middle 50% of the data.

4. **Whiskers**: The lines extending from the top and bottom of each box (the "whiskers") show the range of the data, typically to 1.5 times the IQR above the third quartile and below the first quartile. Points outside of this range are considered outliers.

5. **Outliers**: Individual points above and below the whiskers indicate outliers, which are data points that fall outside of the typical range (more than 1.5 times the IQR).

6. **Comparison**: The plot allows for a comparison between the number of installs for Free and Paid apps.
   - Free apps have a larger number of installs overall, with a higher median and more variability.
   - Paid apps have fewer installs, as indicated by the lower median and fewer points (installs) in the higher range.

From this graph, we can infer that free apps tend to have a larger number of installs compared to paid apps. The spread and outliers in the free apps category also suggest that while most free apps have a moderate number of downloads, there are some free apps with a very high number of downloads, which are not as common among paid apps.

In [70]:
box = px.box(data_frame=df_apps_clean,
                 y='Installs',
                 x='Type',
                 color='Type',
                 notched=True,    # use notched (separate boxes) as style
                 points='all',    # have all points (normal and outliers in graph)
                 title='How Many Downloads are Paid Apps Giving Up?')

box.update_layout(yaxis=dict(type='log'))

box.show()

# Plotly Box Plots: Revenue by App Category

**Challenge**: See if you can generate the chart below:

<img src=https://imgur.com/v4CiNqX.png>

Looking at the hover text, how much does the median app earn in the Tools category? If developing an Android app costs $30,000 or thereabouts, does the average photography app recoup its development costs?

Hint: I've used 'min ascending' to sort the categories.

In [72]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']

box = px.box(df_paid_apps,
              x='Category',
              y='Revenue Estimate',
              title='How Much Can Paid Apps Earn?')

box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Ballpark Revenue',
                  xaxis={'categoryorder':'min ascending'},
                  yaxis=dict(type='log'))


box.show()

### Outliers Explanation

Outliers are typically observations in your data that don't follow the pattern
of the majority of your data points. They are unusual values, much smaller or much larger than the rest. Think of them like kids on a seesaw that are much heavier or lighter than the others—it throws the balance off.

In the context of box plots:

- The **100th percentile** would indeed be the maximum value if we consider all data points, including outliers. However, in a box plot, the top whisker usually doesn't go all the way to the 100th percentile. Instead, it extends to a value that's considered the "maximum" within a certain boundary, often 1.5 times the interquartile range (IQR) above the 75th percentile (the top of the box).
- Any data points that are beyond this boundary are plotted individually as **outliers**. This doesn't mean they are errors or bad data; they're just significantly higher or lower than the bulk of the data.

In [76]:
df_paid_apps = df_apps_clean[df_apps_clean['Type'] == 'Paid']

In [78]:
box = px.box(df_paid_apps,
              x='Category',
              y='Revenue Estimate',
              title='How Much Can Paid Apps Earn?')

box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Ballpark Revenue',
                  xaxis={'categoryorder':'median ascending'}, # sort by ascending median (includes outliers)
                  yaxis=dict(type='log'))


box.show()

# How Much Can You Charge? Examine Paid App Pricing Strategies by Category

**Challenge**: What is the median price price for a paid app? Then compare pricing by category by creating another box plot. But this time examine the prices (instead of the revenue estimates) of the paid apps. I recommend using `{categoryorder':'max descending'}` to sort the categories.

In [79]:
df_paid_apps.Price.median()

2.99

Some categories have higher median prices than others. This time we see that Medical apps have the most expensive apps as well as a median price of $5.49. In contrast, Personalisation apps are quite cheap on average at `$1.49`. Other categories which higher median prices are Business `($4.99)` and Dating `($6.99)`. It seems like customers who shop in these categories are not so concerned about paying a bit extra for their apps.

In [80]:
box = px.box(df_paid_apps,
                 x='Category',
                 y="Price",
                 title='Price per Category')

box.update_layout(xaxis_title='Category',
                  yaxis_title='Paid App Price',
                  xaxis={'categoryorder':'max descending'},
                  yaxis=dict(type='log'))

box.show()

# Learning Points & Summary

In this lesson we looked at how to:

* Pull a random sample from a DataFrame using .sample()

* How to find duplicate entries with .duplicated() and .drop_duplicates()

* How to convert string and object data types into numbers with .to_numeric()

* How to use plotly to generate beautiful pie, donut, and bar charts as well as box and scatter plots


